### ISSUE TITLE
Coding standard should state how to name test classes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7032]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)| @lvh reported|
|-|-|
|Trac ID|trac#7032|
|Type|enhancement|
|Created|2014-03-14 15:35:36Z|
|Branch|https://github.com/twisted/twisted/tree/document-test-class-names-7032|

IIUC, all of the following exist in Twisted:

 - FooTests
 - FooTestCase
 - TestFoo

I really don't care which one it is, but we should pick one, and possibly file tickets for all the ones that don't match, once we've picked one.

Attachments:

* [doc_7032.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3a/b3a1ea10ad82aaa8caae9e06625d42656993ae71/870e2be6b35b89cbfe1ef056b2893ebfd55fdf8a.patch) (1717 bytes) - added by eeshangarg on 2014-08-21 12:23:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7032 7032
type__enhancement enhancement
reporter__lvh lvh
priority__normal normal
milestone__ 
branch__branches_document_test_class_names_7032 branches/document-test-class-names-7032
branch_author__khorn khorn
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation_easy documentation easy
time__1394811336000000 1394811336000000
changetime__1410739684057016 1410739684057016
version__None None
owner__khorn khorn

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

My preference is ```FooTests```

Here's an example of a recently added test module which follows that style 
 * [twisted/python/test/test_tzhelper.py](twisted/python/test/test_tzhelper.py)

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)|@eeshangarg set owner to @eeshangarg|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_3"></a>|@eeshangarg commented|
|-|-|

Hi!

Here is what I think:

FooTestCase: There are currently 305 test classes that use this style, and there are three code examples in the Testing Standard itself that use this style.

TestFoo: There are 254 test classes that currently use this style, and this style of naming test classes does not appear in any kind of documentation(.rst).

FooTests: There are 209 test classes that use this style, and there are two code examples in the Testing Standard that use this style.

So, according to these numbers, I think it would be easier to modify the name of the test classes that use FooTests and TestFoo to use FooTestCase. And I think we will need to file separate tickets for the tests for each individual component of Twisted.

Also, one last thing, I think the code examples in the Testing Standard should also use the style that we choose to use.

I am really looking forward to working on this, please let me know what you think. :)

Thanks & regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Please use `FooTests`.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_5"></a>|@lvh commented|
|-|-|

+1; please use ```FooTests```. While the numbers do point out a "winner" they don't mention how old that relative code is. I think we all use ```FooTests``` now.

It should(?!) be fairly easy to do this with a regex over the codebase (right?) in which case there's little difference between changing any two to the remaining, blessed style.

Separate tickets for each component sounds fine to me. Also means that we don't block for a long time because this one thing somewhere depends on a particular test having a particular name (it shouldn't; but I wouldn't be surprised if a few examples pop out of the woodwork).

Thanks for working on this, Eeshan!

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_6"></a>|@eeshangarg set owner to @lvh|
|-|-|

Hi! I just submitted a patch which adds guidelines to the Testing standard about how to name test classes. I also modified the code examples in the testing standard to use the new style. I asked on #twisted-dev and I was advised that the "Test Implementation Guidelines" section of the testing standard would be an appropriate place for the new guidelines.

Please let me know if further changes are needed and I'll make them right away.

Thanks for letting me contribute to my favorite project!

Regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Thanks.  It would also be nice if this rule was codified in twistedchecker (hardly anyone reads the full contribution guidelines but it's easy to point people at a tool that will find their style mistakes automatically).

This is not at all a blocker for resolving this ticket but it would be a nice enhancement related to this change.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_8"></a>|@eeshangarg commented|
|-|-|

Hi! I just filed a new ticket for tests in twisted.words. I have already submitted a patch for it and I have also added the review keyword. My patch changes all tests in twisted.words to use FooTests now. I will try to submit patches for all of the other tests as soon as I can. :)

[#7622](https://github.com/twisted/twisted/issues/7622)

Thanks!
Eeshan

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn set owner to @khorn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_10"></a>|@khorn commented|
|-|-|

(In [43073]) Branching to document-test-class-names-7032.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_11"></a>|@khorn commented|
|-|-|

(In [43074]) Apply contributed patch to document the naming of test classes in Twisted.
Refs [#7032](https://github.com/twisted/twisted/issues/7032)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_12"></a>|@khorn commented|
|-|-|

This looks good.  Merging.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_13"></a>|@khorn set status to `closed`|
|-|-|

(In [43075]) Merge document-test-class-names-7032: Document the recommended naming of
test classes in Twisted.

Author: eeshangarg
Reviewers: khorn
Fixes: [#7032](https://github.com/twisted/twisted/issues/7032)

Add some documentation to the Testing Standard which specifies the correct
way of naming test classes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Also from the coding standard:

> Documentation should be formatted with a single sentence or clause per line.

(which perhaps reveals how worthwhile it is to add tons of niggling little gotchas like this to the coding standard).
